prodtype attribute not handled properly in rationale element

When using prodtype="" filter in rationale element, the generated datastream keeps multiple instance of the rationale element without filtering it. E.g. when using:
```xml
[...]
<rationale prodtype="rhel7">
...
</rationale>
<rationale prodtype="debian8">
...
</rationale>
```
Will try to find out why but if you know exactly where the prodtype attribute is handled i would be happy to have some little help

not enought, the error is still here. Here is the bug:
```
[ 83%] [rhel7-tables] generating HTML table for cis references
runtime error: file /home/phil/Travail/scap/scap-security-guide/shared/transforms/shared_xccdf2table-byref.xslt line 195 element copy
Attribute nodes must be added before any child nodes to an element.
error: file /home/phil/Travail/scap/scap-security-guide/build/ssg-rhel7-xccdf.xml
xsltRunStylesheet : run failed
RHEL/7/CMakeFiles/generate-rhel7-table-by-ref-cis.dir/build.make:61 : la recette pour la cible « tables/table-rhel7-cisrefs.html » a échouée
make[3]: *** [tables/table-rhel7-cisrefs.html] Erreur 11
CMakeFiles/Makefile2:20570 : la recette pour la cible « RHEL/7/CMakeFiles/generate-rhel7-table-by-ref-cis.dir/all » a échouée
make[2]: *** [RHEL/7/CMakeFiles/generate-rhel7-table-by-ref-cis.dir/all] Erreur 2
CMakeFiles/Makefile2:20227 : la recette pour la cible « RHEL/7/CMakeFiles/rhel7.dir/rule » a échouée
make[1]: *** [RHEL/7/CMakeFiles/rhel7.dir/rule] Erreur 2
Makefile:6556 : la recette pour la cible « rhel7 » a échouée
make: *** [rhel7] Erreur 2
```
when patching the following:
```diff
--- a/shared/xccdf/services/obsolete.xml
+++ b/shared/xccdf/services/obsolete.xml
@@ -181,16 +181,20 @@ Removing the <tt>telnet-server</tt> package decreases the risk of the telnet ser
-<Rule id="package_telnet_removed" severity="low" prodtype="rhel7">
+<Rule id="package_telnet_removed" severity="low" prodtype="rhel7,debian8">
 <title>Remove telnet Clients</title>
 <description>The telnet client allows users to start connections to other 
 systems via the telnet protocol.</description>
-<ocil><yum-macro install="false" package="telnet" /></ocil>
-<rationale>The <tt>telnet</tt> protocol is insecure and unencrypted. The use
+<ocil prodtype="rhel7"><yum-macro install="false" package="telnet" /></ocil>
+<rationale prodtype="rhel7">The <tt>telnet</tt> protocol is insecure and unencrypted. The use
 of an unencrypted transmission medium could allow an unauthorized user
 to steal credentials. The <tt>ssh</tt> package provides an
 encrypted session and stronger security and is included in Red Hat
 Enterprise Linux.</rationale>
+<rationale prodtype="debian8">The <tt>telnet</tt> protocol is insecure and unencrypted. The use
+of an unencrypted transmission medium could allow an unauthorized user
+to steal credentials. The <tt>ssh</tt> package provides an
+encrypted session and stronger security and is Debian and Ubuntu.</rationale>
 <ident prodtype="rhel7" cce="27305-2" />
 <oval id="package_telnet_removed" />
 <ref cis="2.1.2" cui="3.1.13" />
```
the ocil prodtype usage seems to be okay but not the rationale.
I've updated the shorthand2xccdf.xslt file where i found what seems to be a bug:
```diff
diff --git a/shared/transforms/shared_shorthand2xccdf.xslt b/shared/transforms/shared_shorthand2xccdf.xslt
index cb290838a..58f4b633d 100644
--- a/shared/transforms/shared_shorthand2xccdf.xslt
+++ b/shared/transforms/shared_shorthand2xccdf.xslt
@@ -105,7 +105,7 @@
               <xsl:apply-templates select="rationale"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="rationale[not(rationale/@prodtype)]"/>
+              <xsl:apply-templates select="rationale[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
 ```
but the error is still the same, the rhel7 ds file still contains the two rationale contents instead of only one.

--- a/shared/transforms/shared_shorthand2xccdf.xslt
+++ b/shared/transforms/shared_shorthand2xccdf.xslt
@@ -70,7 +70,7 @@
           </xsl:if>
           <xsl:choose>
             <xsl:when test="contains(title/@prodtype, $prod_type) or title/@prodtype = 'all'">
-              <xsl:apply-templates select="title"/>
+              <xsl:apply-templates select="title[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="title[not(@prodtype)]"/>
@@ -78,7 +78,7 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(description/@prodtype, $prod_type) or description/@prodtype = 'all'">
-              <xsl:apply-templates select="description"/>
+              <xsl:apply-templates select="description[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="description[not(@prodtype)]"/>
@@ -86,7 +86,7 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(warning/@prodtype, $prod_type) or warning/@prodtype = 'all'">
-              <xsl:apply-templates select="warning"/>
+              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type or @prodtype = 'all')]"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="warning[not(@prodtype)]"/>
@@ -94,7 +94,7 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(ref/@prodtype, $prod_type) or ref/@prodtype = 'all'">
-              <xsl:apply-templates select="ref"/>
+              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type or @prodtype = 'all')]"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ref[not(@prodtype)]"/>
@@ -102,39 +102,39 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(rationale/@prodtype, $prod_type) or rationale/@prodtype = 'all'">
-              <xsl:apply-templates select="rationale"/>
+              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="rationale[not(rationale/@prodtype)]"/>
+              <xsl:apply-templates select="rationale[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(requires/@prodtype, $prod_type) or requires/@prodtype = 'all'">
-              <xsl:apply-templates select="requires"/>
+              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="requires[not(requires/@prodtype)]"/>
+              <xsl:apply-templates select="requires[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(conflicts/@prodtype, $prod_type) or conflicts/@prodtype = 'all'">
-              <xsl:apply-templates select="conflicts"/>
+              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="conflicts[not(conflicts/@prodtype)]"/>
+              <xsl:apply-templates select="conflicts[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(platform/@prodtype, $prod_type) or platform/@prodtype = 'all'">
-              <xsl:apply-templates select="platform"/>
+              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="platform[not(platform/@prodtype)]"/>
+              <xsl:apply-templates select="platform[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(ident/@prodtype, $prod_type) or ident/@prodtype = 'all'">
-              <xsl:apply-templates select="ident"/>
+              <xsl:apply-templates select="ident[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ident[not(@prodtype)]"/>
@@ -143,7 +143,7 @@
           <!-- order oval (shorthand tag) first, to indicate to tools to prefer its automated checks -->
           <xsl:choose>
             <xsl:when test="contains(oval/@prodtype, $prod_type) or oval/@prodtype = 'all'">
-              <xsl:apply-templates select="oval"/>
+              <xsl:apply-templates select="oval[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="oval[not(@prodtype)]"/>
@@ -189,7 +189,7 @@
           <xsl:apply-templates select="@*" />
           <xsl:choose>
             <xsl:when test="contains(title/@prodtype, $prod_type) or title/@prodtype = 'all'">
-              <xsl:apply-templates select="title"/>
+              <xsl:apply-templates select="title[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="title[not(@prodtype)]"/>
@@ -197,7 +197,7 @@
           </xsl:choose>
            <xsl:choose>
             <xsl:when test="contains(description/@prodtype, $prod_type) or description/@prodtype = 'all'">
-              <xsl:apply-templates select="description"/>
+              <xsl:apply-templates select="description[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="description[not(@prodtype)]"/>
@@ -205,7 +205,7 @@
           </xsl:choose>
            <xsl:choose>
             <xsl:when test="contains(warning/@prodtype, $prod_type) or warning/@prodtype = 'all'">
-              <xsl:apply-templates select="warning"/>
+              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="warning[not(@prodtype)]"/>
@@ -213,7 +213,7 @@
           </xsl:choose>
            <xsl:choose>
             <xsl:when test="contains(ref/@prodtype, $prod_type) or ref/@prodtype = 'all'">
-              <xsl:apply-templates select="ref"/>
+              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ref[not(@prodtype)]"/>
@@ -221,7 +221,7 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(rationale/@prodtype, $prod_type) or rationale/@prodtype = 'all'">
-              <xsl:apply-templates select="rationale"/>
+              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="rationale[not(@prodtype)]"/>
@@ -229,23 +229,23 @@
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(requires/@prodtype, $prod_type) or requires/@prodtype = 'all'">
-              <xsl:apply-templates select="requires"/>
+              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="requires[not(requires/@prodtype)]"/>
+              <xsl:apply-templates select="requires[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(conflicts/@prodtype, $prod_type) or conflicts/@prodtype = 'all'">
-              <xsl:apply-templates select="conflicts"/>
+              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
-              <xsl:apply-templates select="conflicts[not(conflicts/@prodtype)]"/>
+              <xsl:apply-templates select="conflicts[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
             <xsl:when test="contains(platform/@prodtype, $prod_type) or platform/@prodtype = 'all'">
-              <xsl:apply-templates select="platform"/>
+              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="platform[not(@prodtype)]"/>

--- a/shared/transforms/shared_shorthand2xccdf.xslt
+++ b/shared/transforms/shared_shorthand2xccdf.xslt
@@ -69,72 +69,99 @@
             </xsl:attribute>
           </xsl:if>
           <xsl:choose>
-            <xsl:when test="contains(title/@prodtype, $prod_type) or title/@prodtype = 'all'">
-              <xsl:apply-templates select="title[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(title/@prodtype, $prod_type)">
+              <xsl:apply-templates select="title[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="title/@prodtype = 'all'">
+              <xsl:apply-templates select="title[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="title[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(description/@prodtype, $prod_type) or description/@prodtype = 'all'">
-              <xsl:apply-templates select="description[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(description/@prodtype, $prod_type)">
+              <xsl:apply-templates select="description[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="description/@prodtype = 'all'">
+              <xsl:apply-templates select="description[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="description[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(warning/@prodtype, $prod_type) or warning/@prodtype = 'all'">
-              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type or @prodtype = 'all')]"/>
+            <xsl:when test="contains(warning/@prodtype, $prod_type)">
+              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="warning/@prodtype = 'all'">
+              <xsl:apply-templates select="warning[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="warning[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(ref/@prodtype, $prod_type) or ref/@prodtype = 'all'">
-              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type or @prodtype = 'all')]"/>
+            <xsl:when test="contains(ref/@prodtype, $prod_type)">
+              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="ref/@prodtype = 'all'">
+              <xsl:apply-templates select="ref[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ref[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(rationale/@prodtype, $prod_type) or rationale/@prodtype = 'all'">
-              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(rationale/@prodtype, $prod_type)">
+              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="rationale/@prodtype = 'all'">
+              <xsl:apply-templates select="rationale[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="rationale[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(requires/@prodtype, $prod_type) or requires/@prodtype = 'all'">
-              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(requires/@prodtype, $prod_type)">
+              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="requires/@prodtype = 'all'">
+              <xsl:apply-templates select="requires[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="requires[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(conflicts/@prodtype, $prod_type) or conflicts/@prodtype = 'all'">
-              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(conflicts/@prodtype, $prod_type)">
+              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="conflicts/@prodtype = 'all'">
+              <xsl:apply-templates select="conflicts[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="conflicts[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(platform/@prodtype, $prod_type) or platform/@prodtype = 'all'">
-              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(platform/@prodtype, $prod_type)">
+              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="platform/@prodtype = 'all'">
+              <xsl:apply-templates select="platform[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="platform[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(ident/@prodtype, $prod_type) or ident/@prodtype = 'all'">
-              <xsl:apply-templates select="ident[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(ident/@prodtype, $prod_type)">
+              <xsl:apply-templates select="ident[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="ident/@prodtype = 'all'">
+              <xsl:apply-templates select="ident[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ident[not(@prodtype)]"/>
@@ -142,8 +169,11 @@
           </xsl:choose>
           <!-- order oval (shorthand tag) first, to indicate to tools to prefer its automated checks -->
           <xsl:choose>
-            <xsl:when test="contains(oval/@prodtype, $prod_type) or oval/@prodtype = 'all'">
-              <xsl:apply-templates select="oval[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(oval/@prodtype, $prod_type)">
+              <xsl:apply-templates select="oval[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="oval/@prodtype = 'all'">
+              <xsl:apply-templates select="oval[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="oval[not(@prodtype)]"/>
@@ -188,64 +218,88 @@
         <Group>
           <xsl:apply-templates select="@*" />
           <xsl:choose>
-            <xsl:when test="contains(title/@prodtype, $prod_type) or title/@prodtype = 'all'">
-              <xsl:apply-templates select="title[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(title/@prodtype, $prod_type)">
+              <xsl:apply-templates select="title[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="title/@prodtype = 'all'">
+              <xsl:apply-templates select="title[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="title[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
            <xsl:choose>
-            <xsl:when test="contains(description/@prodtype, $prod_type) or description/@prodtype = 'all'">
-              <xsl:apply-templates select="description[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(description/@prodtype, $prod_type)">
+              <xsl:apply-templates select="description[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="description/@prodtype = 'all'">
+              <xsl:apply-templates select="description[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="description[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
            <xsl:choose>
-            <xsl:when test="contains(warning/@prodtype, $prod_type) or warning/@prodtype = 'all'">
-              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(warning/@prodtype, $prod_type)">
+              <xsl:apply-templates select="warning[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="warning/@prodtype = 'all'">
+              <xsl:apply-templates select="warning[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="warning[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
            <xsl:choose>
-            <xsl:when test="contains(ref/@prodtype, $prod_type) or ref/@prodtype = 'all'">
-              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(ref/@prodtype, $prod_type)">
+              <xsl:apply-templates select="ref[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="ref/@prodtype = 'all'">
+              <xsl:apply-templates select="ref[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="ref[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(rationale/@prodtype, $prod_type) or rationale/@prodtype = 'all'">
-              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(rationale/@prodtype, $prod_type)">
+              <xsl:apply-templates select="rationale[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="rationale/@prodtype = 'all'">
+              <xsl:apply-templates select="rationale[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="rationale[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(requires/@prodtype, $prod_type) or requires/@prodtype = 'all'">
-              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(requires/@prodtype, $prod_type)">
+              <xsl:apply-templates select="requires[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="requires/@prodtype = 'all'">
+              <xsl:apply-templates select="requires[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="requires[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(conflicts/@prodtype, $prod_type) or conflicts/@prodtype = 'all'">
-              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(conflicts/@prodtype, $prod_type)">
+              <xsl:apply-templates select="conflicts[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="conflicts/@prodtype = 'all'">
+              <xsl:apply-templates select="conflicts[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="conflicts[not(@prodtype)]"/>
             </xsl:otherwise>
           </xsl:choose>
           <xsl:choose>
-            <xsl:when test="contains(platform/@prodtype, $prod_type) or platform/@prodtype = 'all'">
-              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type) or @prodtype = 'all']"/>
+            <xsl:when test="contains(platform/@prodtype, $prod_type)">
+              <xsl:apply-templates select="platform[contains(@prodtype, $prod_type)]"/>
+            </xsl:when>
+            <xsl:when test="platform/@prodtype = 'all'">
+              <xsl:apply-templates select="platform[@prodtype = 'all']"/>
             </xsl:when>
             <xsl:otherwise>
               <xsl:apply-templates select="platform[not(@prodtype)]"/>

